File: python/datatypes/CostSet.py
from dataclasses import dataclass
from typing import FrozenSet
from .WeightedItem import WeightedItem


@dataclass(frozen=True)
class CostSet:
    """
    A CostSet is a set of WeightedItems characterized by a cost.
    The weight of a CostSet is the sum of the weights of the WeightedItems it contains.
    """

    s: FrozenSet[WeightedItem]
    cost: int
    weight: int = 0

    def __post_init__(self):
        object.__setattr__(self, 'weight', sum(map(lambda x: x.weight, self.s)))

    def __repr__(self):
        return f'CS: (s={self.s}, cost={self.cost}, weight={self.weight})'

File: scala/src/main/scala/datatypes/CostSet.scala
package datatypes

final case class CostSet(s: List[WeightedItem], cost: Int) extends Ordered[CostSet] {
  lazy val weight: Int = s.map(_.weight).sum

  override def toString: String = s"CS(s=$s, c=$cost, w=$weight)"
  
  def compare(that: CostSet): Int = {
    if (this.weight == that.weight) {
      this.cost compare that.cost
    } else if (this.cost == that.cost) {
      that.weight compare this.weight
    } else {
      val tmp = that.weight compare this.weight
      if (tmp == 0) {
        this.cost compare that.cost
      } else {
        tmp
      }
    }
  }
}